[2단계 - 경로 조회 기능] 도이(유도영) 미션 제출합니다.

yoondgu opened the pull request:
안녕하세요 코즈! 🐭  잘 지내셨나요?
2단계 기능 구현을 완료하여 리뷰 요청 드립니다.

구현 하는 과정에서 답을 못내린 부분, 고민해보고 싶은 부분들을 적은 TODO 주석이 여전히 꽤 많이 남아있습니다. 😔  아직 많이 부족하지만 코즈의 의견을 들어보며 진행하고 싶어서 리뷰 요청 드려요! 

주석이 여기 저기 있어 지저분한데요.. 해당 주석 부분에 코멘트로 질문드리도록 하겠습니다!! 코멘트를 남기지 않은 부분은 피드백 반영하며 함께 개선해나가도록 하겠습니다.

1단계 피드백 반영과 추가로 진행한 리팩터링 목록을 [README.md](https://github.com/woowacourse/jwp-subway-path/blob/5acbbcb9c4aa79454db762c0522828ebf5a9659c/README.md#1%EB%8B%A8%EA%B3%84)에 링크와 함께 정리하였습니다.

많이 부족한 만큼 따끔한 리뷰 부탁드리겠습니다!! 감사합니다 😊

---- 

#### 1단계 리팩터링 관련

> 도이가 이해하는 도메인 모델은 어떤 방식인가요?
추상화된 모델로서 이해하는 방식이 모든 정보를 삭제 후 저장하는 방식이라면 그에 맞게 구현하는것이 좋다고 생각합니다. 성능상 문제가 발생하면, 그때 어떻게 해결할 수 있을지 고민할 것 같아요.

이 말씀을 듣고 보니, 이 프로그램에서 도메인 모델은 어떤 역할을 해야할지에 대한 생각이 명확해졌어요!

그리고 API 명세는 언제든지 변경될 수 있는 요소이기 때문에 이를 통해 비즈니스 로직을 단순화(가능하더라도)하려는 시도는 부적절할 수 있다고 생각하게 되었어요.

이상적인 도메인 모델은 그 자체로 이 프로그램에서 나타내는 도메인의 요구사항을 모두 표현하고 수행할 수 있어야 한다고 생각합니다. 컨트롤러에서 클라이언트와 어떻게 통신할지, DAO에서 DB에 어떻게 액세스할지에 대해서 최대한 영향 없이 독립적으로 존재하는 것이 유지보수성이 좋고, 유연한 프로그램이 될 수 있을 것 같아요.
변경사항에 대해 DB에서 모두 삭제/저장하는 비효율 에 대한 문제는 별도 고려사항인 것 같습니다. (변경을 감지하는 별도 로직을 만드는 등, 추후 성능 문제 발생 시 고려해도 되는 문제)

이에 따라 서비스 계층에서는 최대한 도메인과 DB를 연결하는 역할만 하고, 도메인에서 핵심 로직을 담당하도록 전면 리팩터링하였습니다.

- 기존의 Subway라는 객체가 노선의 역 연결 정보를 조회하는 역할만 담당했는데, 이 객체에서 역 등록/삭제를 한 뒤 적절한 Section 목록을 반환하도록 변경했습니다.
- 전체적으로 클래스명을 보다 적절하다고 생각하는 이름으로 바꾸고, [도메인 별 요구사항](https://github.com/woowacourse/jwp-subway-path/blob/5acbbcb9c4aa79454db762c0522828ebf5a9659c/README.md#%EB%8F%84%EB%A9%94%EC%9D%B8-%EC%9A%94%EA%B5%AC%EC%82%AC%ED%95%AD)을 다시 정리해보았습니다.

yoondgu committed:
refactor: 도메인 객체 자체로 요구사항을 만족하도록 전면 리팩터링

- Subway 객체가 한 노선의 역 연결 정보를 위한 로직을 직접 수행하도록 변경
- 노선의 경로를 표현하는 도메인 객체 네이밍 수정 Subway -> LineRoute
- 역 연결을 표현하는 그래프 생성 로직 팩터리 클래스로 분리
- 서비스 계층에서는 도메인과 DB를 연결만 하도록 변경
- 전면 리팩터링에 따라 사용하지 않는 메서드 삭제
- 객체 책임 분리
    - Section 클래스로 구간 분할, 병합 책임 이동
    - Distance 클래스로 거리 계산 책임 이동

yoondgu committed:
refactor: 역 연결 그래프 생성 메서드 검증로직 추가

yoondgu committed:
feat: 구간 서비스 검증 로직 추가, 테스트 작성, 트랜잭션 적용

yoondgu committed:
fix: 노선 역이 두 개일 때 하나만 삭제되는 버그 수정, 테스트 작성

yoondgu committed:
docs: 리팩터링 목록에 코드 리뷰 피드백 추가

yoondgu committed:
feat: 예측하지 못한 예외 정보 콘솔 출력 대신 로그 사용하기

yoondgu committed:
refactor: 테스트를 위한 메서드 삭제

yoondgu committed:
docs: 코멘트 답변한 1단계 피드백 목록 체크 업데이트

yoondgu committed:
refactor: 데이터 초기화 Controller 대신 sql 파일 사용

yoondgu committed:
refactor: 객체 id로 동일성, 동등성 판단하도록 수정

yoondgu committed:
refactor: 노선 경로 객체에서 불필요한 노선 객체 참조 삭제

yoondgu committed:
test: 불필요한 필드 삭제

yoondgu committed:
refactor: 역 연결 그래프 커스텀 클래스로 변경, 팩토리 메서드 생성자 사용 방지

yoondgu committed:
refactor: 사용하지 않는 생성자 삭제

yoondgu committed:
feat: 커스텀 예외 정의, 예외 처리 기능 추가

yoondgu committed:
feat: DTO null 검증 로직 추가

yoondgu committed:
refactor: 노선 별 역 연결 정보 클래스명 변경

- 경로 조회 요구사항 추가에 따라 헷갈리지 않도록 이름 변경함
- LineRoute -> LineMap

yoondgu committed:
docs: 2단계 요구사항 추가에 따른 미션 명세서 업데이트

yoondgu committed:
test: 구간 테스트 픽스쳐 상수명 구체적으로 수정

yoondgu committed:
refactor: 불필요한 와일드카드 제거

yoondgu committed:
feat: 모든 노선 역 연결 그래프 생성을 위한 팩토리 구현

yoondgu committed:
feat: 최단 경로 조회 API 설계

yoondgu committed:
refactor: 노선 역 등록 및 삭제 API 설계 변경

- 특정 노선에 대한 API이므로 /lines의 하위 URI로 변경함

yoondgu committed:
refactor: 최단 경로 조회 API 설계 보완, 서비스 일부 구현

yoondgu committed:
refactor: 모든 노선 역 연결 그래프 생성 시 정상적인 구간 정보만 허용하도록 수정

- 그래프 생성 시 팩토리 패턴 대신 정적 팩토리 메서드 패턴 사용하여 검증 방식 강제함
- 모든 노선 역 연결 그래프 생성 시 노선 역 연결 그래프를 전달받도록 함
- 이를 통해 정상적인 구간 정보만 허용하도록 함

yoondgu committed:
refactor: 지하철 노선도 객체 최단 거리 경로 찾는 기능 구현

yoondgu committed:
refactor: 최단 거리 경로에 대한 총 거리, 요금 계산 구현

yoondgu committed:
refactor: 지하철 노선도 경로 탐색 시 검증 로직 추가

yoondgu committed:
refactor: 경로 서비스 핸들러 메서드 분리

yoondgu committed:
test: 테스트 픽스쳐 관련 수정

- 노선 별 구간 목록 픽스쳐 정의

yoondgu committed:
test: 경로 서비스 테스트 코드 작성

yoondgu committed:
fix: 에러 응답 DTO getter 정의

yoondgu committed:
feat: 프로덕션 DB 로컬 저장 설정

yoondgu committed:
fix: 테스트 더미데이터 DDL 오작동 수정

- 데이터가 계속 새로 들어가는 문제 해결.
- id로 데이터를 식별하므로 해당 행의 id도 지정해 insert함.

yoondgu committed:
chore: 실행 시 편의를 위해 프로덕션 db 더미데이터 추가

yoondgu committed:
fix: 경로 조회 관련 누락된 예외 검증 로직 추가

- 사용자에 의한 에러: 출발 역, 도착 역이 동일한 경우
- 서버 문제에 의한 에러: DB에 한 노선에 대한 중복 구간이 존재하는 경 (비정상적인 역 연결 정보 생성)
- 예외 추가에 따라 일관성을 위해 기존 예외 클래스명 수정

yoondgu committed:
test: 경로 조회 통합테스트 잘못된 요청 케이스 작성

yoondgu committed:
refactor: 테스트를 위한 메서드 삭제

yoondgu committed:
refactor: 역 등록 관련 테스트 케이스 추가

yoondgu committed:
fix: 빌드 설정 관계없이 오류 안나도록 DTO에 기본 생성자 추가

- Intellij IDEA로 빌드 시 HttpMessageConversionException 발생하는 문제 해결

yoondgu committed:
test: 테스트 클래스 displayName 작성, fixture 패키지 분리

yoondgu committed:
refactor: 노선 상태를 가진 그래프 변 클래스명 변경

yoondgu committed:
test: 사용하지 않는 테스트 삭제

yoondgu committed:
build: 도메인 패키지 구조 세분화

- 아래와 같은 기준으로 분류함
- entity: 도메인 엔티티 또는 테이블 엔티티 역할을 하는 클래스 (현재는 구분없이 사용중)
- vo: 값 객체의 역할만 하는 클래스
- 그 외: 엔티티, 값 객체를 사용하여 더 복잡한 로직을 수행하는 도메인 클래스

yoondgu committed:
chore: properties 파일 불필요한 빈줄 제거

yoondgu committed:
refactor: 모호한 예외 클래스명 수정

- 클라이언트 측의 원인으로 발생한 예외라는 의미로 네이밍 변경

yoondgu committed:
build: 프로덕션 코드 구간 더미데이터 DDL 삭제

- 구간 데이터(SECTIONS 테이블)는 프로그램 사용에 따라 변하기 때문에 insert ignore문 사용이 부적절
- 어플리케이션 실행 시마다 기존 DDL 사용하면 잘못된 데이터 값이 들어가게 됨
- 따라서 해당 더미데이터 기본 실행되지 않도록 삭제함

yoondgu committed:
refactor: 클래스명 자동 변경으로 바뀐 이름 복구

yoondgu commented on src/main/java/subway/application/RouteService.java:
```diff
@@ -0,0 +1,72 @@
+        Station sourceStation = stationDao.findById(sourceStationId)
+                .orElseThrow(() -> new RequestDataNotFoundException("출발 역이 존재하지 않습니다."));
+        Station targetStation = stationDao.findById(targetStationId)
+                .orElseThrow(() -> new RequestDataNotFoundException("도착 역이 존재하지 않습니다."));
```
전달받은 역의 id를 통해 DB에서 역을 가져오고, 없을 경우 예외를 발생하는 로직이
RouteService, LineService 두 서비스에서 사용됩니다.
하지만 RouteService가 LineService를 참조하기에는 어색한 관계라고 생각되어 고민입니다.
어떻게 보면 stationDao의 findById라는 메서드가 이미 재사용 가능하게 만들어져 있기 때문에,
이런 중복로직이 있는 것은 당연한 걸까요..? 코즈의 조언을 듣고 싶어요!!

kouz95 replied:
RouteService가 LineService를 의존했을때 발생할 수 있는 문제는 무엇이 있나요?

kouz95 replied:
RouteService가 LineService를 의존했을때 발생할 수 있는 문제는 무엇이 있나요?

yoondgu replied:
앗 제가 질문, 주석을 잘못 썼네요. LineService가 아닌 SectionService를 이야기하려고 했습니다. 혼란을 드려 죄송해요 😓  
그런데 다시 보니 SectionDao, StationDao 대신 SectionService를 의존한다고 해서 중복 로직을 해결하는 상황은 아니었던 것 같습니다.  
서비스 클래스의 분리를 잘 못한 건지 현재로서는 불가피한 중복이라고 생각되어 그대로 유지하였습니다..!  ㅜㅜ  

다만 만약 중복 로직을 해결하는 상황이었다고 했을 때  서비스 간의 의존 관계가 문제가 된다고 생각했던 이유는,  
같은 Dao 코드를 사용하더라도 A와 B가 서로 연관 없는 비즈니스 로직을 수행하고 있다면  
A가 B를 의존하게 했을 때, B의 변경에 A가 영향을 받을 수 있기 때문입니다.

kouz95 replied:
> A가 B를 의존하게 했을 때, B의 변경에 A가 영향을 받을 수 있기 때문입니다.   

더 핵심적인 영역에 있는 객체의 변경을 전파 받는것은 필요한 일입니다. 도메인 영역 (핵심 영역)에서의 변경이 발생하면 그 변경을 의존하고 있는 모든 영역에서 반영해주는 것이 바람직합니다.  
`orElseThrow` 까지 하는 메서드를 dao나 Service에 public 메서드로 만들어주고 그 메서드를 다른 서비스에서 사용하는 것이 큰 문제를 만들진 않을 것이라 생각해요 :)

yoondgu committed:
refactor: 모호한 메서드명 변경

yoondgu commented on src/main/java/subway/domain/RoutedStations.java:
```diff
@@ -0,0 +1,121 @@
+// TODO 노선 별 요금 적용을 위해 StationEdge로 변경, 생성할 때 Line 전달받아 Edge에 저장하기 or Section이 Line을 가지면 해결되는 문제 아닌가?
+// Section이 Line을 가지면, Line이 사실상 비즈니스 로직에 필요 없을 때에도 Line을 저장해야 한다 (노선 역 등록/삭제)
+// 그래프에서 필요에 따라 edge에 Line을 저장하게 만들자. => 필요에 따라 Line을 가지고 있는 edge, Line을 가지고 있지 않은 edge를 사용하도록 함
+// StationEdge -> LinedClassifiableWeightedEdge
```
1. 외부 라이브러리와의 강한 결합
외부 라이브러리에서 제공받는 내용을 노출시키지 않고, 이를 사용하는 방식/생성 방식을 강제하고 싶어서 위와 같이 상속한 클래스를 사용하고 있습니다.  
하지만 라이브러리 교체가 어렵고 너무 외부 라이브러리 의존적인 코드가 된 것 같습니다.    
인터페이스를 적용하는 방식으로 변경하는 게 좋을까요?  

2. 구간을 노선 별로 구분하기

사실 처음에는 3단계 요구사항 중 노선 별 요금 정책 또한 고려하며 구현하려고 했는데요..ㅎㅎ  
구간을 노선 별로 구분할 수 있어야 하는 요구사항이 생기더라도  
Section은 Line을 모르고, 외부 객체가 둘의 관계를 관리하는 기존 방향을 유지하고 싶었습니다.  

그래서 `MultiRoutedStations`객체에서는 어느 Section이 어느 Line인지 관리하도록 하기 위해서
그래프의 `LineClassifiableWeightedEdge`가 Line을 상태로 가지도록 만들어보았습니다.

하지만 막상 만들고 나니, 이 클래스에서는 어떤 edge를 사용해야 할지 애매해져서 일단 3단계 고려는 보류하고 기본 구현을 진행했습니다.

애매해진 이유는 이 클래스는 원래 "노선 역 연결 정보"로 만든건데, "최단 경로 역 연결 정보"를 전달할 때에도 쓰이기 때문입니다.  
노선 역 연결 정보는 모두 같은 노선이므로 Line 정보가 필요 없는 반면, 노선 별 요금 정책을 추가했을 때 최단 경로 역 연결 정보에는 각 구간의 Line 정보가 필요하게 될 것입니다. 
(`List<Section>` 그대로 전달하면 제대로 연결된 구간 목록인지 알 수 없으므로 그래프로 변환해 반환해야 한다고 생각했습니다.)

따라서 이 방향으로 진행하려면
"최단 경로 역 연결 정보"를 담기 위해, 이 클래스와 거의 같지만 `LineClassifiableWeightedEdge`를 사용하는 그래프 클래스를 따로 만들어 사용해야 할 것 같습니다.


하지만 Section이 Line에 대해 알면 이 모든 고민들이 필요 없는 것 같기도 하고
제가 생각지 못한 다른 방법을 두고 비효율적으로 설계하고 있는 것은 아닌지,
불필요하게 복잡도를 높이고 있는지 고민이 됩니다..!

kouz95 replied:
1. 이 부분은 취향일수도 있을것 같아요. 전 다형성이 아니라 단치 구체적인 클래스를 의존하는 것을 막기 위한 DIP는 개인적으로 선호하지 않습니다. Testable한 구조를 만들기 위해서 DIP를 적용하는것이 좋다고 주장하는 경우도 있고, 전부 DIP를 적용하면서 작업을 해보기도 했는데, 코드가 복잡해지는 것 같아서요. 다형성으로 풀어내기에 적절한 상황이 아니면 그냥 의존시키는 것 같아요.  

2. 복잡성을 늘리는 원인이 무엇이라고 생각하시나요?  
`Section은 Line을 모르고, 외부 객체가 둘의 관계를 관리하는 기존 방향`을 유지하려다보니 발생하는 건가요? Section이 Line에 대해 알면, 지금 복잡하다고 느끼는 부분이 해결되나요?

yoondgu replied:
Section - Line 관계가 문제가 아니고 서로 다른 목적에 대해 같은 자료구조를 쓰는 것이 문제였던 것 같아서  
경로 조회에 쓰이는 자료구조(TransferableRoute)를 별도로 정의하였습니다!

yoondgu commented on src/main/java/subway/domain/SubwayMap.java:
```diff
@@ -0,0 +1,54 @@
+
+    public RoutedStations findShortestRoutedStations(final Station sourceStation, final Station targetStation) {
+        // TODO 외부 라이브러리 자체적으로 비정상적인 경우에 예외를 던지는데, 그 예외를 전환시키기 vs 그와 상관없이 미리 모두 검증하기
+        // 외부 라이브러리에서 던져주는 예외에 의존하면, 추후 라이브러리 변경 시 어려움이 클 것 같다.
```
저는 주석에 작성된 이유를 근거로 후자를 선택해 진행했는데요.  
그럼에도 제가 예측하지 못한 외부 라이브러리의 예외가 발생할 것 같다는 불안감이 계속 있는 것 같습니다..  
- 외부 라이브러리에서 발생한 예측하지 못한 예외를 이 메서드에서 try-catch로 잡아 적절한 예외 및 메시지로 전환하는 게 좋을까요?  
모든 예외를 잡을 수는 없으므로 ExceptionHandler에서 서버 에러로 처리하는 것만으로도 충분하다고 생각하기는 했는데요.  
외부 라이브러리의 예외를 핸들링하는 저만의 기준이 아직 명확하지 않은 것 같습니다.    
- 혹시 코즈는 어떤 기준을 가지고 계신지도 궁금합니다..!!

kouz95 replied:
코드로 명시해주는 것이 좋은 것 같습니다.  
여러 사람이 관리하게 되는 코드라면, 라이브러리 안에서 검증이 된다는 히스토리를 남기는 것 보단, 코드로 표시되어있는것이 더 명확한것 같아요.  
라이브러리에서 체크하도록 한다했을때의 장점은 잘 안떠올라요

yoondgu commented on src/test/java/subway/application/RouteServiceTest.java:
```diff
@@ -0,0 +1,99 @@
+        RouteResponse result = routeService.findShortestRoute(1L, 9L);
+
+        // TODO 협력객체에 대한 의존을 피하고자 했는데, 결국 Dao 외에 도메인 협력객체에 대해서는 의존하여 테스트를 하고 있는 것이 아닌가?
+        // TODO 이렇게까지 서비스에서 테스트해야할까?
```
저는 협력 객체에 대한 단위 테스트가 별도로 잘 이루어지고 있다면(여기서는 Dao 테스트, 도메인 객체 별 단위 테스트)  
테스트의 목적에 따라 가짜 객체를 충분히 사용할 수 있다고 생각합니다.
또, 제가 생각하는 서비스 테스트의 목적은 DB-도메인 간 조율을 잘 하고 있는지 "행위"를 검증하는 것이라고 생각해 Stubbing을 사용했습니다.  
그래서 Dao에 대한 Stubbing은 세부 구현이 모두 드러난다는 단점 외에는 현재 문제가 없다고 생각하는데,
아래 코드에서는 도메인 협력객체를 직접 사용해서 의도와 다른 것 같아 고민입니다.

아래처럼 값을 계산하지말고, 기대되는 값을 직접 적어주는 게 맞을까요? (expectedTotalFare 대신 18)   
아니면 서비스에서 도메인이 해야할 일(요금 계산)을 하고 있어 발생한 문제일까요?

kouz95 replied:
calculate를 FareCalculator 의 정적 메서드로 분리한 이유는 무엇인가요?  
현재 사용처는 RouteService 한 곳인거 같은데, RouteService 의 메서드 였을 수도 있지 않나요?  
만약 RouteService 에 계산식이 존재했다면, 테스트 코드는 어떻게 작성했을 것 같나요?

yoondgu replied:
1. 요금 계산에 대해서 다른 도메인 협력 객체는 책임이 무관하고, RouteService가 알기에는 구체적인 정책이라 생각해 별도의 클래스로 분리했습니다.  
- 그런데 요금 계산을 하는 객체는 상태를 가질 필요가 없고, 2단계에서는 요금을 교체할 필요가 없기 때문에 정적 메서드만 사용하도록 만들었습니다.  
- 그렇지만 한 곳에서만 쓰이는데 정적 메서드로 만들었더니, 불필요한 남용이 일어나는 것 같아요 (제가 테스트코드에서 사용한 것처럼요)   그리고 클래스만 나뉘었지 사실상 RouteService가 요금 계산을 직접 책임지는 상황이라 제가 생각한 도메인-서비스 계층의 분리가 잘 이루어지지 않은 것 같습니다.  
- 따라서 RoutedStations 객체가 요금 정책을 전달받아 요금을 계산하도록 수정하려고 합니다.   

  
2. 만약 RouteService에 계산식이 존재했다면, 테스트 코드에서도 직접 동일한 계산식을 사용해 기대값을 구한 뒤 결과와 비교했을 것 같습니다.  
- 지금은 FareCalculator의 정적 메서드는 어디에서나 접근 가능하기 때문에 그냥 이를 써버렸고 그래서 의도와는 다른 방식으로 테스트하고 있는 것 같아요. ㅎㅎ  
- 하지만 이와는 별개로 사실 이 서비스 테스트에서 거리 계산, 요금 계산의 결과값을 확인하는 것 자체가 본래 의도와 다른 것 같아요.   FareCalculator, Distance에서 해당하는 부분에 대한 테스트를 하고 있기 때문에 이 테스트 코드에서는 아예 삭제하려고 합니다.  
혹시 이에 대해 코즈는 어떻게 생각하시나요..?

kouz95 replied:
1, 2번에 대해 어떻게 생각하는지 질문 주신걸까용?  

> RouteService가 알기에는 구체적인 정책이라 생각해   

구체적인 정책과 그렇지 않은 정책은 어떻게 구별할 수 있나요? 좀 더 명확하고 정량적인 기준을 세워볼 순 없을까요?  

> 테스트 코드에서는 아예 삭제하려고 합니다.  
 
Service에서 의존하고 있는 다른 객체들의 동작까지 테스트하고 있기 때문에 다른 테스트와 중복된 테스트가 되었다는 생각이 드셨던것 같아요. Mockito의 verify 메서드 등을 통해 서비스에서 호출이 일어났는지만 테스트 했었더라면 중복이 없는 테스트이었을 수도 있을것 같아요.  

Service에선 호출만 테스트해야한다고 말씀드리는 것은 아닙니다. 테스트엔 여러 종류가 있기도 하고 (단위 테스트, 통합 테스트, 인수 테스트 등..) 각각 테스트를 통해 얻고자 하는 가치가 다르다고 생각합니다. 때문에, 유지보수성을 해치지 않는 선에선 중복된 영역을 테스트해도 문제가 되진 않는다고 생각해요

yoondgu commented on src/test/java/subway/domain/SubwayMapTest.java:
```diff
@@ -0,0 +1,89 @@
+@DisplayName("지하철 노선도 단위 테스트")
+class SubwayMapTest {
+
+    // TODO 어떻게 하면 다양한 테스트 케이스를 확인할 수 있을까?
```
복잡한 요구사항에 대해 편하게 테스트하기 위해 Test fixture를 적극적으로 사용하게 되었습니다.  
테스트하기 좋은 케이스(환승 가능, 거리가 다른 경로)를 fixture로 미리 설정해두니 편하지만  
fixture 간에 의존관계가 생기고, 한 테스트에서 사용하는 fixture를 변경했을 때 다른 테스트도 영향을 받게 됩니다.  
(LINE1_SECTIONS의 내용을 바꾸면, 다른 테스트가 실패)  

이런 문제가 없으려면 불편하더라도 fixture를 사용하지 않거나
지금처럼 fixture 클래스를 두지 말고 각 테스트 클래스에서 상수로 관리하는 방법 뿐일까요?

kouz95 replied:
LINE1_SECTIONS 라는 이름 보다, 무엇을 연출하기 위한 객체인지 이름으로 관리되면 좋을것 같아요.  
예를 들어, LINE1_SECTIONS 가 아니라, `연속적으로_다수_연결된_호선` 과 같은 이름으로 관리가 된다면, 내용을 바꾸더라도 예상치 못한 실패가 발생하는 일은 적어지지 않을까요? (갑자기 비 연속적인 SECTION을 추가하는 경우엔 실패하는 테스트가 생기는게 정상일 것 같아요)

yoondgu replied:
테스트 픽스쳐에서는 네이밍으로 의도를 드러내는 게 중요할 수 있다는 것을 알게 되었어요!! 감사합니다 ㅎㅎ  
연속적으로 연결되어있고, 어떤 호선에서 환승이 가능하도록 만들어두었는지 `ROUTED_SECTIONS_1_TRANSFER_2_AT_ST1`와 같이 명시하였습니다!

yoondgu commented:
안녕하세요 코즈! 🐭  잘 지내셨나요?
2단계 기능 구현을 완료하여 리뷰 요청 드립니다.

구현 하는 과정에서 답을 못내린 부분, 고민해보고 싶은 부분들을 적은 TODO 주석이 여전히 꽤 많이 남아있습니다. 😔  아직 많이 부족하지만 코즈의 의견을 들어보며 진행하고 싶어서 리뷰 요청 드려요! 

주석이 여기 저기 있어 지저분한데요.. 해당 주석 부분에 코멘트로 질문드리도록 하겠습니다!! 코멘트를 남기지 않은 부분은 피드백 반영하며 함께 개선해나가도록 하겠습니다.

1단계 피드백 반영과 추가로 진행한 리팩터링 목록을 [README.md](https://github.com/woowacourse/jwp-subway-path/blob/5acbbcb9c4aa79454db762c0522828ebf5a9659c/README.md#1%EB%8B%A8%EA%B3%84)에 링크와 함께 정리하였습니다.

많이 부족한 만큼 따끔한 리뷰 부탁드리겠습니다!! 감사합니다 😊

---- 

#### 1단계 리팩터링 관련

> 도이가 이해하는 도메인 모델은 어떤 방식인가요?
추상화된 모델로서 이해하는 방식이 모든 정보를 삭제 후 저장하는 방식이라면 그에 맞게 구현하는것이 좋다고 생각합니다. 성능상 문제가 발생하면, 그때 어떻게 해결할 수 있을지 고민할 것 같아요.

이 말씀을 듣고 보니, 이 프로그램에서 도메인 모델은 어떤 역할을 해야할지에 대한 생각이 명확해졌어요!

그리고 API 명세는 언제든지 변경될 수 있는 요소이기 때문에 이를 통해 비즈니스 로직을 단순화(가능하더라도)하려는 시도는 부적절할 수 있다고 생각하게 되었어요.

이상적인 도메인 모델은 그 자체로 이 프로그램에서 나타내는 도메인의 요구사항을 모두 표현하고 수행할 수 있어야 한다고 생각합니다. 컨트롤러에서 클라이언트와 어떻게 통신할지, DAO에서 DB에 어떻게 액세스할지에 대해서 최대한 영향 없이 독립적으로 존재하는 것이 유지보수성이 좋고, 유연한 프로그램이 될 수 있을 것 같아요.
변경사항에 대해 DB에서 모두 삭제/저장하는 비효율 에 대한 문제는 별도 고려사항인 것 같습니다. (변경을 감지하는 별도 로직을 만드는 등, 추후 성능 문제 발생 시 고려해도 되는 문제)

이에 따라 서비스 계층에서는 최대한 도메인과 DB를 연결하는 역할만 하고, 도메인에서 핵심 로직을 담당하도록 전면 리팩터링하였습니다.

- 기존의 Subway라는 객체가 노선의 역 연결 정보를 조회하는 역할만 담당했는데, 이 객체에서 역 등록/삭제를 한 뒤 적절한 Section 목록을 반환하도록 변경했습니다.
- 전체적으로 클래스명을 보다 적절하다고 생각하는 이름으로 바꾸고, [도메인 별 요구사항](https://github.com/woowacourse/jwp-subway-path/blob/5acbbcb9c4aa79454db762c0522828ebf5a9659c/README.md#%EB%8F%84%EB%A9%94%EC%9D%B8-%EC%9A%94%EA%B5%AC%EC%82%AC%ED%95%AD)을 다시 정리해보았습니다.

kouz95 commented on README.md:
```diff
@@ -74,11 +136,76 @@ HTTP/1.1 201
+    - [ ] API 문서 자동화 도구 사용하기
+
+- 피드백
+    - [x] [LineRequest, LineResponse와 같이 외부 영역에서 정의되는 자료구조를 서비스에서 사용하면 어떤 문제가 있을까?](https://github.com/woowacourse/jwp-subway-path/pull/78#discussion_r1192853316)
```
발생하는 문제의 원인과 해결 방안들에 대해서 잘 고민해주신것 같습니다  
모든 계층과 행위마다 분리하는 방식은 이후 발생할수 있는 모든 상황들에 대응 가능한 구조이긴 하지만, 생산성이 현저히 떨어져서요. 전 문제가 생길때 분리하는 방식을 선호 합니다 :)  
핵심 계층이 (재사용성이 높은 계층이) 외부 계층 (클라이언트 계층)의 자료구조 혹은 관심사를 의존하게 될때 문제가 발생할 수 있다는 점을 인지해주시고, 실제로 문제가 발생할때의 해결법만 인지하면 좋을것 같아요

kouz95 replied:
발생하는 문제의 원인과 해결 방안들에 대해서 잘 고민해주신것 같습니다
모든 계층과 행위마다 분리하는 방식은 이후 발생할수 있는 모든 상황들에 대응 가능한 구조이긴 하지만, 생산성이 현저히 떨어져서요. 전 문제가 생길때 분리하는 방식을 선호 합니다 :)
핵심 계층이 (재사용성이 높은 계층이) 외부 계층 (클라이언트 계층)의 자료구조 혹은 관심사를 의존하게 될때 문제가 발생할 수 있다는 점을 인지해주시고, 실제로 문제가 발생할때의 해결법만 인지하면 좋을것 같아요

kouz95 commented on src/main/java/subway/application/LineService.java:
```diff
@@ -43,16 +44,15 @@ public LineResponse findLineResponseById(Long id) {
-        Subway subway = Subway.of(persistLine, sections);
-        List<Station> orderedStations = subway.getOrderedStations();
-        List<StationResponse> stationResponses = orderedStations.stream()
+    private List<StationResponse> convertToOrderedStationResponses(final List<Section> sections) {
```
`StationResponse`에 `List<Section>`을 받아 `List<StationResponse>`을 만들어내는 정적 팩터리 메서드를 만들어줄 수 도 있을것 같네요.  
Service가 Response를 만드는 책임을 갖는 것이 당장 문제가 생기는 부분은 아니지만, 변환 코드가 계속해서 늘어나게 되는 구조일 것 같아서요

kouz95 commented on src/main/java/subway/application/LineService.java:
```diff
@@ -43,16 +44,15 @@ public LineResponse findLineResponseById(Long id) {
-        Subway subway = Subway.of(persistLine, sections);
-        List<Station> orderedStations = subway.getOrderedStations();
-        List<StationResponse> stationResponses = orderedStations.stream()
+    private List<StationResponse> convertToOrderedStationResponses(final List<Section> sections) {
```
`StationResponse`에 `List<Section>`을 받아 `List<StationResponse>`을 만들어내는 정적 팩터리 메서드를 만들어줄 수 도 있을것 같네요.  
Service가 Response를 만드는 책임을 갖는 것이 당장 문제가 생기는 부분은 아니지만, 변환 코드가 계속해서 늘어나게 되는 구조일 것 같아서요

yoondgu replied:
그러고 보니 변환 코드들이 이곳저곳 계속 늘어나서 코드를 파악하기 더 어려웠던 것 같아요!
이 부분을 포함해 변환 로직들을 해당하는 DTO의 정적 팩터리 메서드로 바꾸어 주었습니다.

그런데 이때, `List<Section>->List<Station>->List<StationResponse>` 이렇게 여러 번의 변환을 해야 하는 상황에서
`StationResponse`가 `Station`보다 상위 개념인 `Section`을 알고 있는 게 좋지 않다고 생각이 들었습니다.
Section과 Station의 관계가 달라지거나 이를 가져오는 로직이 달라지면 함께 변경되어야 하기 때문입니다.

반면 `List<Station>->List<StationResponse>` 이 부분은 여러 번 재사용되는 상황이기 때문에
`Section->List<Station>`의 변환은 도메인 객체의 책임으로 분리하고
`List<Station>`을 받아 `List<StationResponse>`을 만드는 정적 팩터리 메서드를 만들었습니다.

kouz95 commented on src/main/java/subway/application/LineService.java:
```diff
@@ -43,16 +44,15 @@ public LineResponse findLineResponseById(Long id) {
-        return stationResponses;
     }
 
     public Line findLineById(Long id) {
```
요 메서드는 private으로 접근제한자가 바뀌어도 될것 같네요 :)

yoondgu replied:
그러네요..!! 수정하겠습니다 🙂

kouz95 commented on src/main/java/subway/application/RouteService.java:
```diff
@@ -0,0 +1,72 @@
+        return getRouteResponse(sourceStation, targetStation, sectionsByLine);
+    }
+
+    private RouteResponse getRouteResponse(final Station sourceStation, final Station targetStation,
```
이 부분들도 Reponse의 정적 팩터리 메서드로 처리되면 어떨까요?

kouz95 reviewed (changes requested):
안녕하세요 도이 :)  
리뷰 확인해주세요

yoondgu committed:
refactor: 메서드 접근제한자 수정, 변환 로직 정적 팩터리 메서드로 변경

- private 이어도 되는 메서드 접근제한자 수정
- 부적절한 정적 팩터리 메서드 네이밍 수정
- 변환 로직은 해당 DTO의 정적 팩터리 메서드로 만들어 관리가 쉽도록 수정
- 역 연결 정보 관련 책임 이동

yoondgu committed:
test: 서비스 테스트 코드의 불필요한 검증 삭제

yoondgu committed:
refactor: 외부 라이브러리에 대한 예측 못한 예외 처리 코드로 명시

yoondgu committed:
refactor: 환승 가능 경로 조회 클래스 분리

yoondgu committed:
refactor: 요금 계산 클래스 분리
- SubwayMap 에서 요금 정책과 TransferableRoute 를 전달받아 계산하도록 변경
- 요금 정책 변경을 쉽게 하기 위하여 인터페이스 적용

yoondgu committed:
chore: 피드백 답변 업데이트 및 TODO 주석 삭제

yoondgu committed:
refactor: 테스트 픽스쳐 네이밍에 사용 목적 명시

yoondgu commented:
안녕하세요 코즈! 피드백 반영해보았습니다.
3단계를 고려하며 피드백 반영해보려고 했는데요. (FareCalculator 인터페이스 부분),
고민하다보니 정작 시간이 부족해서 .. 3단계까지 마저 진행하면 시간 내에 요청을 못 드릴 것 같았어요.
2단계 피드백 반영까지만 하고 리뷰 요청드립니다. 🥲 
이번에도 잘 부탁드려요. 감사합니다!

kouz95 reviewed (approved):
안녕하세요 도이 :)  
이번 단계에서 요구하는 수준은 충분히 만족하였다고 판단하여 머지합니다.  
고생하셨습니다
